core/common/src/main/java/alluxio/network/thrift/PeekableTransport.java

```diff
@@ -0,0 +1,131 @@
+/*
+ * The Alluxio Open Foundation licenses this work under the Apache License, version 2.0
+ * (the "License"). You may not use this work except in compliance with the License, which is
+ * available at www.apache.org/licenses/LICENSE-2.0
+ *
+ * This software is distributed on an "AS IS" basis, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
+ * either express or implied, as more fully set forth in the License.
+ *
+ * See the NOTICE file distributed with this work for information regarding copyright ownership.
+ */
+
+package alluxio.network.thrift;
+
+import org.apache.thrift.transport.TTransport;
+import org.apache.thrift.transport.TTransportException;
+
+import com.google.common.base.Preconditions;
+
+/**
+ * A transport that one can peek a given number of bytes from the read stream without changing the
+ * read position.
+ */
+public class PeekableTransport extends TTransport {
+  private TTransport mUnderlyingTransport;
+  private byte[] mBuffer;
```

Comment:
could this be improved with `ByteBuffer`?

Reply:
I used `System.arraycopy` in this implementation for performance, which works better with arrays.

Reply:
ack